### ISSUE TITLE
Offloading .d.ts generation to typeshare

### ISSUE THREAD

**Issue by martinjlowm** [napi-rs/napi-rs#2086]
There seems to be a fair bit of functional overlap with [Typeshare](https://github.com/1Password/typeshare) in terms of emitting type declarations from Rust types. Typeshare scopes a broader set of target languages and currently only support structs and enums. 

One benefit typeshare has is that it doesn't seem to depend on macro expansion when emitting the declaration file (in typeshare's perspective it's really just a ts file, but the interpretation is mostly the same), which skips Rust compilation. That detail is what got me to write up this issue in the first place. We have a dedicated GitHub Action workflow for type checking and we'd like to generate the declaration file without checking it into source control - in fact, it's not a single project but rather tens of projects. We would spend a significant amount of (unnecessary I'd argue) time building with napi just to generate a declaration file.

Have any considerations been made to offload the .d.ts-file generation to the typeshare project by contributing the parts that it lacks which this project has, e.g. traits, assuming they are interested in extending their features? It should be noted that both projects use `syn` to parse Rust, so the parsing implementations look familiar.

**Comment by alexeagle** [comment]
Similarly under Bazel, I already have a JS loader and binding generation via the `rust_shared_library` function. I ONLY need a tool to produce the typings, and want to skip the rest of what `napi build` is doing.

**Comment by Brooooooklyn** [comment]
See https://napi.rs/docs/cli/programmatic-api#standalone-typesjs-generation

**Comment by martinjlowm** [comment]
I'll have a look!

**Comment by martinjlowm** [comment]
It is not quite what I was hoping for. The `typeDefDir` still needs to be generated and that is the thing that takes a long time.

**Comment by martinjlowm** [comment]
@alexeagle - I opened up https://github.com/napi-rs/napi-rs/pull/3124 a few days ago if you want to give it a spin

**Comment by alloveras** [comment]
> Similarly under Bazel, I already have a JS loader and binding generation via the rust_shared_library function.

Yes, we have a similar setup internally and we struggled with this as well (the fact that the types are emitted on an un-declared directory that isn't captured nor propagated as part of `rust_shared_library`).

Our (arguably hacky) solution was to extend `napi-derive` (via patch) with an additional `embed-type-defs` feature that embeds each `#[napi]`-annotated export's serialised type definition as a fixed-size byte static in a dedicated linker section of the compiled shared library.

This way, the output of `rust_shared_library` keeps being a single shared object (`.so` or `.dylib`) but with the type information now embedded into it (using the "distributed slice" linker-section pattern). So, in the Bazel context, we can write a tiny rule that takes that shared object as an input, reads that section in one pass and produces the same intermediate JSON file that`@napi-rs/cli` expects.

If this implementation sounds interesting, I'd be happy to contribute it as a small pull request that is purely additive to `napi-derive` crate.
